kernel/arch/dreamcast/util/screenshot.c

```diff
@@ -24,53 +28,88 @@ This will now work with any of the supported video modes.
 */
 
 int vid_screen_shot(const char *destfn) {
-    file_t  f;
-    int i, numpix;
-    uint8   *buffer;
-    char    header[256];
-    uint32  save;
-    uint32  pixel;  /* to fit 888 mode */
-    uint8   r, g, b;
-    uint8   bpp;
+    file_t   f;
+    uint8_t  *buffer;
+    char     header[256];
+    int      i, numpix;
+    uint32_t save;
+    uint32_t pixel, pixel1, pixel2;  /* to fit 888 mode */
+    uint8_t  r, g, b;
+    uint8_t  bpp;
 
     bpp = 3;    /* output to ppm is 3 bytes per pixel */
     numpix = vid_mode->width * vid_mode->height;
 
     /* Allocate a new buffer so we can blast it all at once */
-    buffer = (uint8 *)malloc(numpix * bpp);
+    buffer = (uint8_t *)malloc(numpix * bpp);
 
     if(buffer == NULL) {
         dbglog(DBG_ERROR, "vid_screen_shot: can't allocate ss memory\n");
         return -1;
     }
 
+    /* Open output file */
+    f = fs_open(destfn, O_WRONLY | O_TRUNC);
+
+    if(!f) {
+        dbglog(DBG_ERROR, "vid_screen_shot: can't open output file '%s'\n", destfn);
+        free(buffer);
+        return -1;
+    }
+
     /* Disable interrupts */
     save = irq_disable();
 
-    /* Write out each pixel as 24 bits */
+    /* Write out each pixel as 24-bits */
     switch(vid_mode->pm) {
         case(PM_RGB555): {
-            for(i = 0; i < numpix; i++) {
-                pixel = vram_s[i];
-                r = (((pixel >> 10) & 0x1f) << 3);
-                g = (((pixel >>  5) & 0x1f) << 3);
-                b = (((pixel >>  0) & 0x1f) << 3);
-                buffer[i * 3 + 0] = r;
-                buffer[i * 3 + 1] = g;
-                buffer[i * 3 + 2] = b;
+            /* Process two 16-bit pixels at a time */
+            for(i = 0; i < numpix/2; i++) {
+                pixel = vram_l[i];
+                pixel1 = pixel & 0xFFFF;
+                pixel2 = pixel >> 16;
+
+                /* Process the first pixel */
+                r = (((pixel1 >> 10) & 0x1f) << 3);
+                g = (((pixel1 >> 5) & 0x1f) << 3);
+                b = (((pixel1 >> 0) & 0x1f) << 3);
+                buffer[i * 6 + 0] = r;
+                buffer[i * 6 + 1] = g;
+                buffer[i * 6 + 2] = b;
+
+                /* Process the second pixel */
+                r = (((pixel2 >> 10) & 0x1f) << 3);
+                g = (((pixel2 >> 5) & 0x1f) << 3);
+                b = (((pixel2 >> 0) & 0x1f) << 3);
+                buffer[i * 6 + 3] = r;
+                buffer[i * 6 + 4] = g;
+                buffer[i * 6 + 5] = b;
```

Comment:
You can actually go even faster by processing 4 pixels at a time. Then you read two 32-bit words, and you write three 32-bit words:
```
READ:  0AAAAABBBBBCCCCC0DDDDDEEEEEFFFFF 0GGGGGHHHHHIIIII0JJJJJKKKKKLLLLL
WRITE: AAAAA000BBBBB000CCCCC000DDDDD000 EEEEE000FFFFF000GGGGG000HHHHH000 IIIII000JJJJJ000KKKKK000LLLLL000
```

So when processing 4 pixels, instead of doing twelve 8-bit writes, you only do three 32-bit writes.

Reply:
Oh...durr. I used the same approach here: https://github.com/KallistiOS/KallistiOS/pull/576/files and didnt even thing of applying it here.  Thanks for the feedback. Im on it.

Reply:
I did am implementation of what you said and it was around ~2 milliseconds slower.  What am I doing wrong?  Here is my code:

```       
#define EXTRACT_RED(pixel)   ((((pixel) >> 10) & 0x1F) << 3)
#define EXTRACT_GREEN(pixel) ((((pixel) >> 5) & 0x1F) << 3)
#define EXTRACT_BLUE(pixel)  (((pixel) & 0x1F) << 3)

#define RGB555_COMPOSE_RGBR(pixel1, pixel2) \
    (((uint32_t)EXTRACT_RED(pixel2) << 24) | \
    ((uint32_t)EXTRACT_BLUE(pixel1) << 16) | \
    ((uint32_t)EXTRACT_GREEN(pixel1) << 8) | \
    EXTRACT_RED(pixel1))

#define RGB555_COMPOSE_GBRG(pixel1, pixel2) \
    (((uint32_t)EXTRACT_GREEN(pixel2) << 24) | \
    ((uint32_t)EXTRACT_RED(pixel2) << 16) | \
    ((uint32_t)EXTRACT_BLUE(pixel1) << 8) | \
    EXTRACT_GREEN(pixel1))

#define RGB555_COMPOSE_BRGB(pixel1, pixel2) \
    (((uint32_t)EXTRACT_BLUE(pixel2) << 24) | \
    ((uint32_t)EXTRACT_GREEN(pixel2) << 16) | \
    ((uint32_t)EXTRACT_RED(pixel2) << 8) | \
    EXTRACT_BLUE(pixel1))
            
            vram_ll = (uint64_t *)vram_l;
            buffer32 = (uint32_t *)buffer;

            uint64_t start = timer_ms_gettime64();
            /* Process four 16-bit pixels at a time */
            for(i = 0; i < numpix/4; i++) {
                pixel64 = vram_ll[i];
                pixel1 = pixel64 & 0xFFFF;
                pixel2 = (pixel64 >> 16) & 0xFFFF;
                pixel3 = (pixel64 >> 32) & 0xFFFF;
                pixel4 = (pixel64 >> 48) & 0xFFFF;

                buffer32[i * 3 + 0] = RGB555_COMPOSE_RGBR(pixel1, pixel2);
                buffer32[i * 3 + 1] = RGB555_COMPOSE_GBRG(pixel2, pixel3);
                buffer32[i * 3 + 2] = RGB555_COMPOSE_BRGB(pixel3, pixel4);
            }
            uint64_t end = timer_ms_gettime64();
            printf("Time: %lld\n", end - start);```